dkg/src/transcript_aggregation.rs

```diff
@@ -0,0 +1,176 @@
+        },
+        transcript_bytes: good_trx_bytes.clone(),
+    });
+    assert!(matches!(result, Err(_)));
+
+    // Node with invalid transcript should be rejected.
+    let mut bad_trx_bytes = good_trx_bytes.clone();
+    bad_trx_bytes[0] = 0xAB;
+    let result = trx_agg_state.add(addrs[2], DKGNode {
+        metadata: DKGTranscriptMetadata {
+            epoch: 999,
+            author: addrs[2],
+        },
+        transcript_bytes: vec![],
+    });
+    assert!(matches!(result, Err(_)));
+
+    // Good node should be accepted.
+    let result = trx_agg_state.add(addrs[3], DKGNode {
+        metadata: DKGTranscriptMetadata {
+            epoch: 999,
+            author: addrs[3],
+        },
+        transcript_bytes: good_trx_bytes.clone(),
+    });
+    assert!(matches!(result, Ok(None)));
+
+    // Node from contributed author should be ignored.
+    let result = trx_agg_state.add(addrs[3], DKGNode {
+        metadata: DKGTranscriptMetadata {
+            epoch: 999,
+            author: addrs[3],
+        },
+        transcript_bytes: good_trx_bytes.clone(),
+    });
+    assert!(matches!(result, Ok(None)));
+
+    // Aggregated trx should be returned if after adding a node, the threshold is exceeded.
+    let result = trx_agg_state.add(addrs[4], DKGNode {
+        metadata: DKGTranscriptMetadata {
+            epoch: 999,
+            author: addrs[4],
+        },
+        transcript_bytes: good_trx_bytes.clone(),
+    });
+    assert!(matches!(result, Ok(Some(_))));
+}
+
+impl<S: DKGTrait> BroadcastStatus<DKGMessage> for Arc<TranscriptAggregationState<S>> {
+    type Aggregated = S::Transcript;
+    type Message = DKGNodeRequest;
+    type Response = DKGNode;
+
+    fn add(&self, sender: Author, dkg_node: DKGNode) -> anyhow::Result<Option<Self::Aggregated>> {
+        let DKGNode {
+            metadata,
+            transcript_bytes,
+        } = dkg_node;
+        ensure!(
+            metadata.epoch == self.epoch_state.epoch,
+            "adding dkg node failed with invalid node epoch",
+        );
+        ensure!(
+            metadata.author == sender,
+            "adding dkg node failed with node author mismatch"
+        );
+        let transcript = bcs::from_bytes(transcript_bytes.as_slice())?;
+        let mut trx_aggregator = self.trx_aggregator.lock();
+        if trx_aggregator.contributors.contains(&metadata.author) {
+            return Ok(None);
+        }
+
+        S::verify_transcript(&self.dkg_pub_params, &transcript)?;
+
+        // All checks passed. Aggregating.
+        trx_aggregator.contributors.insert(metadata.author);
+        if let Some(agg_trx) = trx_aggregator.trx.as_mut() {
+            S::aggregate_transcripts(&self.dkg_pub_params, agg_trx, &transcript);
+        } else {
+            trx_aggregator.trx = Some(transcript);
+        }
+        let maybe_aggregated = self
+            .epoch_state
+            .verifier
+            .check_voting_power(trx_aggregator.contributors.iter(), true)
```

Comment:
Note we only need f+1 here. But 2f+1 is also fine as the DKG performance is not that critical.